# import data


In [10]:
import pandas as pd
import numpy as np

train_features = pd.read_csv('https://raw.githubusercontent.com/Sebercheres/Machine_Learning_Final_Project/main/Dataset/X_train.csv', index_col=0)
test_features = pd.read_csv('https://raw.githubusercontent.com/Sebercheres/Machine_Learning_Final_Project/main/Dataset/X_valid.csv', index_col=0)
train_labels = pd.read_csv('https://raw.githubusercontent.com/Sebercheres/Machine_Learning_Final_Project/main/Dataset/y_train.csv', index_col=0)
test_labels = pd.read_csv('https://raw.githubusercontent.com/Sebercheres/Machine_Learning_Final_Project/main/Dataset/y_valid.csv', index_col=0)

In [11]:
!pip3 install catboost

# Baseline Model

In [12]:
from sklearn.dummy import DummyClassifier
from sklearn.metrics import precision_score

dum = DummyClassifier(random_state=7)
dum.fit(train_features, train_labels)
predict = dum.predict(test_features)
prec = precision_score(test_labels, predict,average='macro')
prec

/usr/local/lib/python3.7/dist-packages/sklearn/dummy.py:132: FutureWarning: The default value of strategy will change from stratified to prior in 0.24.
  "stratified to prior in 0.24.", FutureWarning)


0.48973782106312225

# add model

In [15]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import AdaBoostClassifier
from xgboost import XGBClassifier
from catboost import CatBoostClassifier
from lightgbm import LGBMClassifier


from sklearn.model_selection import train_test_split
from sklearn.model_selection import RandomizedSearchCV, GridSearchCV
from sklearn.model_selection import validation_curve
from sklearn.metrics import recall_score, f1_score, precision_score, accuracy_score, roc_auc_score

In [17]:
models =  { "Decision Tree" : DecisionTreeClassifier(),
            "RandomForest" : RandomForestClassifier(random_state=7),
            "XGBoost" : XGBClassifier(), 
            "LogisticRegression" : LogisticRegression(max_iter=500), 
            "CatBoost" :CatBoostClassifier() ,
            'LGBM' : LGBMClassifier(),
            'KNN' : KNeighborsClassifier(),
            'ADA' : AdaBoostClassifier(),

            
            }

scores = pd.DataFrame(columns=['Model','Accuracy','Precision','Recall','F1','ROC'])
probability = {}

for model in models:
  classifier = models[model]
  classifier.fit(train_features, train_labels)
  predicts = classifier.predict(test_features)
  score = classifier.predict_proba(test_features)[:,1]
  roc = roc_auc_score(test_labels, score, average='weighted')
  probability[model] = score

  
  scores = scores.append(pd.Series({
                 'Model' : model,
                 'Accuracy' : accuracy_score(test_labels, predicts),
                 'F1' : f1_score(test_labels, predicts, average ='weighted'),
                 'Precision' : precision_score(test_labels, predicts, average ='weighted'),
                 'Recall' : recall_score(test_labels, predicts, average ='weighted'),
                 'ROC' : roc
  }),ignore_index=True)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
/usr/local/lib/python3.7/dist-packages/sklearn/preprocessing/_label.py:235: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.7/dist-packages/sklearn/preprocessing/_label.py:268: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, 

Learning rate set to 0.015543
0:	learn: 0.6908620	total: 10.7ms	remaining: 10.7s
1:	learn: 0.6896191	total: 23.3ms	remaining: 11.6s
2:	learn: 0.6881399	total: 32.1ms	remaining: 10.7s
3:	learn: 0.6862058	total: 46.9ms	remaining: 11.7s
4:	learn: 0.6841862	total: 57ms	remaining: 11.3s
5:	learn: 0.6828527	total: 70.1ms	remaining: 11.6s
6:	learn: 0.6815550	total: 74ms	remaining: 10.5s
7:	learn: 0.6803443	total: 77.9ms	remaining: 9.66s
8:	learn: 0.6788832	total: 81.8ms	remaining: 9.01s
9:	learn: 0.6777463	total: 85.7ms	remaining: 8.48s
10:	learn: 0.6765091	total: 89.4ms	remaining: 8.04s
11:	learn: 0.6748158	total: 93.2ms	remaining: 7.67s
12:	learn: 0.6736310	total: 96.9ms	remaining: 7.36s
13:	learn: 0.6726582	total: 101ms	remaining: 7.09s
14:	learn: 0.6714999	total: 105ms	remaining: 6.86s
15:	learn: 0.6700994	total: 108ms	remaining: 6.67s
16:	learn: 0.6688461	total: 112ms	remaining: 6.49s
17:	learn: 0.6673021	total: 116ms	remaining: 6.35s
18:	learn: 0.6660106	total: 120ms	remaining: 6.22s
19

/usr/local/lib/python3.7/dist-packages/sklearn/preprocessing/_label.py:235: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.7/dist-packages/sklearn/preprocessing/_label.py:268: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y,

In [19]:
def score_model(model):
  model.fit(train_features, train_labels)
  predicts = model.predict(test_features)
  return accuracy_score(test_labels, predicts)

In [18]:
scores

Model  Accuracy  Precision    Recall        F1       ROC
0       Decision Tree  0.571646   0.582636  0.571646  0.575552  0.562266
1        RandomForest  0.679878   0.676373  0.679878  0.653261  0.705063
2             XGBoost  0.672256   0.687739  0.672256  0.622354  0.685625
3  LogisticRegression  0.611280   0.762614  0.611280  0.465411  0.494834
4            CatBoost  0.690549   0.686601  0.690549  0.669871  0.729131
5                LGBM  0.679878   0.671518  0.679878  0.667904  0.703467
6                 KNN  0.637195   0.624737  0.637195  0.624718  0.649771
7                 ADA  0.649390   0.648898  0.649390  0.594201  0.614692

the best model without parameter tuning was RandomForest so we are going to try tuning it

In [9]:
rf = RandomForestClassifier(random_state = 42)
from pprint import pprint
# Look at parameters used by our current forest
print('Parameters currently in use:\n')
pprint(rf.get_params())

Parameters currently in use:

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'auto',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_impurity_split': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': 42,
 'verbose': 0,
 'warm_start': False}


In [20]:
from sklearn.model_selection import RandomizedSearchCV
# Number of trees in random forest
n_estimators = [int(x) for x in np.linspace(start = 200, stop = 2000, num = 10)]
# Number of features to consider at every split
max_features = ['auto', 'sqrt']
# Maximum number of levels in tree
max_depth = [int(x) for x in np.linspace(10, 110, num = 11)]
max_depth.append(None)
# Minimum number of samples required to split a node
min_samples_split = [2, 5, 10]
# Minimum number of samples required at each leaf node
min_samples_leaf = [1, 2, 4]
# Method of selecting samples for training each tree
bootstrap = [True, False]
# Create the random grid
random_grid = {'n_estimators': n_estimators,
               'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf,
               'bootstrap': bootstrap}
pprint(random_grid)

{'bootstrap': [True, False],
 'max_depth': [10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, None],
 'max_features': ['auto', 'sqrt'],
 'min_samples_leaf': [1, 2, 4],
 'min_samples_split': [2, 5, 10],
 'n_estimators': [200, 400, 600, 800, 1000, 1200, 1400, 1600, 1800, 2000]}


In [21]:
# Use the random grid to search for best hyperparameters
# First create the base model to tune
rf = RandomForestClassifier()
# Random search of parameters, using 3 fold cross validation, 
# search across 100 different combinations, and use all available cores
rf_random = RandomizedSearchCV(estimator = rf, param_distributions = random_grid, n_iter = 100, cv = 3, verbose=2, random_state=42, n_jobs = -1)
# Fit the random search model
rf_random.fit(train_features, train_labels)

Fitting 3 folds for each of 100 candidates, totalling 300 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  37 tasks      | elapsed:  4.1min
[Parallel(n_jobs=-1)]: Done 158 tasks      | elapsed: 17.6min
[Parallel(n_jobs=-1)]: Done 300 out of 300 | elapsed: 31.9min finished
/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_search.py:739: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  self.best_estimator_.fit(X, y, **fit_params)


RandomizedSearchCV(cv=3, error_score=nan,
                   estimator=RandomForestClassifier(bootstrap=True,
                                                    ccp_alpha=0.0,
                                                    class_weight=None,
                                                    criterion='gini',
                                                    max_depth=None,
                                                    max_features='auto',
                                                    max_leaf_nodes=None,
                                                    max_samples=None,
                                                    min_impurity_decrease=0.0,
                                                    min_impurity_split=None,
                                                    min_samples_leaf=1,
                                                    min_samples_split=2,
                                                    min_weight_fraction_leaf=0.0,
               

In [22]:
rf_random.best_params_

{'bootstrap': True,
 'max_depth': 10,
 'max_features': 'auto',
 'min_samples_leaf': 1,
 'min_samples_split': 5,
 'n_estimators': 1600}

In [121]:
base_model = RandomForestClassifier(random_state=42)
score_model(base_model)

0.698170731707317

In [76]:
best_random = RandomForestClassifier(random_state=42, **rf_random.best_params_)
score_model(best_random)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  


0.6814024390243902

In [44]:
from sklearn.model_selection import GridSearchCV
# Create the parameter grid based on the results of random search 
param_grid = {
    'bootstrap': [True],
    'max_depth': [80, 90, 100, 110],
    'max_features': [2, 3],
    'min_samples_leaf': [3, 4, 5],
    'min_samples_split': [8, 10, 12],
    'n_estimators': [100, 200, 300, 1000]
}
# Create a based model
rf = RandomForestClassifier()
# Instantiate the grid search model
grid_search = GridSearchCV(estimator = rf, param_grid = param_grid, 
                          cv = 3, n_jobs = -1, verbose = 2)

In [118]:
# Fit the grid search to the data
grid_search.fit(train_features, train_labels)
grid_search.best_params_

Fitting 3 folds for each of 288 candidates, totalling 864 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  37 tasks      | elapsed:   51.6s
[Parallel(n_jobs=-1)]: Done 158 tasks      | elapsed:  4.0min
[Parallel(n_jobs=-1)]: Done 361 tasks      | elapsed:  9.3min
[Parallel(n_jobs=-1)]: Done 644 tasks      | elapsed: 16.9min
[Parallel(n_jobs=-1)]: Done 864 out of 864 | elapsed: 22.9min finished


{'bootstrap': True,
 'max_depth': 110,
 'max_features': 3,
 'min_samples_leaf': 3,
 'min_samples_split': 10,
 'n_estimators': 200}

In [120]:
best_grid = RandomForestClassifier(random_state=42, **grid_search.best_params_)
score_model(best_grid)

0.6859756097560976

In [114]:
predicts = best_grid.predict(test_features)

In [115]:
Accuracy = accuracy_score(test_labels, predicts)
F1 = f1_score(test_labels, predicts, average ='weighted')
Precision = precision_score(test_labels, predicts, average ='weighted')
Recall = recall_score(test_labels, predicts, average ='weighted')
print(Accuracy, F1, Precision, Recall)

0.6890243902439024 0.6601987401585089 0.690904577331686 0.6890243902439024


In [105]:
CBC = CatBoostClassifier()
parameters = {'depth'         : [4,5,6,7,8,9, 10],
              'learning_rate' : [0.01,0.02,0.03,0.04],
              'iterations'    : [10, 20,30,40,50,60,70,80,90, 100]
              }
Grid_CBC = GridSearchCV(estimator=CBC, param_grid = parameters, cv = 2, n_jobs=-1)
Grid_CBC.fit(train_features, train_label)

Custom logger is already specified. Specify more than one logger at same time is not thread safe.

0:	learn: 0.6893362	total: 30ms	remaining: 2.97s
1:	learn: 0.6861261	total: 45.5ms	remaining: 2.23s
2:	learn: 0.6833429	total: 61.1ms	remaining: 1.97s
3:	learn: 0.6799917	total: 76.9ms	remaining: 1.85s
4:	learn: 0.6770016	total: 94.5ms	remaining: 1.8s
5:	learn: 0.6741814	total: 111ms	remaining: 1.74s
6:	learn: 0.6706956	total: 127ms	remaining: 1.69s
7:	learn: 0.6676408	total: 143ms	remaining: 1.64s
8:	learn: 0.6645027	total: 160ms	remaining: 1.62s
9:	learn: 0.6613717	total: 177ms	remaining: 1.59s
10:	learn: 0.6590263	total: 196ms	remaining: 1.59s
11:	learn: 0.6562800	total: 212ms	remaining: 1.55s
12:	learn: 0.6534163	total: 231ms	remaining: 1.54s
13:	learn: 0.6512947	total: 246ms	remaining: 1.51s
14:	learn: 0.6486056	total: 262ms	remaining: 1.48s
15:	learn: 0.6467077	total: 277ms	remaining: 1.46s
16:	learn: 0.6442373	total: 292ms	remaining: 1.43s
17:	learn: 0.6421455	total: 308ms	remaining: 1.4s
18:	learn: 0.6402424	total: 323ms	remaining: 1.38s
19:	learn: 0.6384729	total: 338ms	remain

GridSearchCV(cv=2, error_score=nan,
             estimator=<catboost.core.CatBoostClassifier object at 0x7fbb3d981890>,
             iid='deprecated', n_jobs=-1,
             param_grid={'depth': [4, 5, 6, 7, 8, 9, 10],
                         'iterations': [10, 20, 30, 40, 50, 60, 70, 80, 90,
                                        100],
                         'learning_rate': [0.01, 0.02, 0.03, 0.04]},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=0)

In [112]:
model = CatBoostClassifier(random_state=42,  **Grid_CBC.best_params_)
score_model(model)

0:	learn: 0.6899207	total: 17.7ms	remaining: 1.75s
1:	learn: 0.6861746	total: 35.9ms	remaining: 1.76s
2:	learn: 0.6827847	total: 53.3ms	remaining: 1.72s
3:	learn: 0.6799309	total: 71.5ms	remaining: 1.72s
4:	learn: 0.6777181	total: 90.3ms	remaining: 1.72s
5:	learn: 0.6749053	total: 108ms	remaining: 1.7s
6:	learn: 0.6715378	total: 128ms	remaining: 1.7s
7:	learn: 0.6690050	total: 145ms	remaining: 1.67s
8:	learn: 0.6663289	total: 162ms	remaining: 1.64s
9:	learn: 0.6631559	total: 181ms	remaining: 1.63s
10:	learn: 0.6606701	total: 200ms	remaining: 1.62s
11:	learn: 0.6581214	total: 223ms	remaining: 1.63s
12:	learn: 0.6557438	total: 241ms	remaining: 1.61s
13:	learn: 0.6532779	total: 258ms	remaining: 1.59s
14:	learn: 0.6509426	total: 276ms	remaining: 1.56s
15:	learn: 0.6487476	total: 293ms	remaining: 1.54s
16:	learn: 0.6466162	total: 311ms	remaining: 1.52s
17:	learn: 0.6442604	total: 330ms	remaining: 1.5s
18:	learn: 0.6417780	total: 348ms	remaining: 1.49s
19:	learn: 0.6391926	total: 366ms	remai

0.6951219512195121

In [116]:
predicts = model.predict(test_features)

In [1]:
from sklearn.metrics import classification_report

y_pred_rf= model.predict(X_test)
print(classification_report(test_labels, y_pred_rf))

NameError: ignored

In [117]:
Accuracy = accuracy_score(test_labels, predicts)
F1 = f1_score(test_labels, predicts, average ='weighted')
Precision = precision_score(test_labels, predicts, average ='weighted')
Recall = recall_score(test_labels, predicts, average ='weighted')
roc_auc_
print(Accuracy, F1, Precision, Recall)

0.6951219512195121 0.6492164807182483 0.7268826666064996 0.6951219512195121


 after testing several models we conclude that catboost preformed the best with accuracy of 69.5% on unseen data which has improved 41% from baseline model